src/ik_memetic.cpp

```diff
@@ -0,0 +1,237 @@
+#include <pick_ik/goal.hpp>
+#include <pick_ik/ik_gradient.hpp>
+#include <pick_ik/ik_memetic.hpp>
+#include <pick_ik/robot.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <fmt/core.h>
+#include <optional>
+#include <vector>
+
+namespace pick_ik {
+
+MemeticIk MemeticIk::from(std::vector<double> const& initial_guess, CostFn const& cost_fn) {
+    return MemeticIk{std::vector<double>(initial_guess.size(), 0.0), cost_fn(initial_guess)};
+}
+
+void MemeticIk::computeExtinctions() {
+    double min_fitness = population_.front().fitness;
+    double max_fitness = population_.back().fitness;
+    for (size_t i = 0; i < population_count_; ++i) {
+        double grading = static_cast<double>(i) / static_cast<double>(population_count_ - 1);
+        population_[i].extinction =
+            (population_[i].fitness + min_fitness * (grading - 1)) / max_fitness;
+    }
+}
+
+void MemeticIk::gradientDescent(size_t const i, Robot const& robot, CostFn const& cost_fn) {
+    auto& individual = population_[i];
+    auto local_ik = GradientIk::from(individual.genes, cost_fn);
+    auto constexpr timeout_local = 0.005;
+    auto constexpr max_iters_local = 25;
```

Comment:
In addition to the step size in the local gradient descent itself, the max time and/or iterations doing this for memetic IK should perhaps also be configured.

What's more, I need to perhaps check if gradient descent has converged BEFORE the max time listed here and return early in that case.